### PR TITLE
Fix some missing code in step-by-step walk through

### LINKED ISSUE #10518: Step-by-step walk-through encounters some error when proceeding with Step-by-Step
## 📚 Documentation

I'm using pytorch-lightning (1.5.1).

When I tried to go through the Step-by-step walk-through of Get Started, From MNIST to AutoEncoders, The Research, to get started with PyTorch Lightning, I found some missing code and an error.

Relevant page: https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#the-research

## 1. lack of print

Affected part: https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#the-model

This is a very small problem, but `Size([1, 10])` requires the `print(out.shape)` statement.

## 2. Missing library

Affected part: https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#data

When I run this, I get the following error.

```
ModuleNotFoundError: No module named 'torchvision'.
```

Solving this is easy, but it might be clear if the following installation command exists in the document.

```
pip install torchvision
```

## 3. lack of import and lack of configure_optimizers method

Affected part: https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#pass-dataloaders-to-fit

### 3.1 Missing import

When I run this, I get the following error.

```
NameError: name 'Trainer' is not defined
```

This is a simple error, but it might be easier to specify import.

```
from pytorch_lightning import Trainer
```

### 3.2 Lack of configure_optimizers method

In addition, even with the above import, I still get the following error.

```
pytorch_lightning.utilities.exceptions.MisconfigurationException: No `configure_optimizers()` method defined. Lightning `Trainer` expects as minimum a `training_step()`, `train_dataloader()` and `configure_optimizers()` to be defined.
````

This is because the class LitMNIST defined above does not have a configure_optimizers method.

The configure_optimizers method is described in the [Optimizer](https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#optimizer) section after here.

So it may be good that we swap the order with [Data](https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#data) section and Optimizer section to solve the problem.

### 3.3 Where similar fixes may be needed

Related to this, it may be also good that we fix code in the beginning of [training](https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#training).

`class LitMNIST(LightningModule)` also appears to need to add a configure_optimizers method.

## 4. Code that I don't know how to execute

Affected part: https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html#datamodules-recommended

Here class MyDataModule(LightningDataModule) is defined, but what is used after that is MNISTDatamodule and ImagenetDatamodule.

It may be desirable to include an example where you can try to run LightningDataModule.

## Fixes these

I don't know the proper solution for "4. Code that I don't know how to execute", but fixing 1, 2, and 3 is easy. So I will create a small pull request to fix 1, 2, and 3.

### PR DESCRIPTION
**Pull request by os1ma**
## What does this PR do?

Fixes #10518

To make it easier for beginners to get started with PyTorch Lightning, I've fixed some errors I encountered in Get Started page.

This pull request is Ready for review.

If there is any problem with this change, please let me know. I'll be happy to fix it!

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by os1ma** [comment]
ci-tests in CircleCI is failing, but this is not because of this change, but because the installation of jsonnet is failing.

```
#!/bin/bash -eo pipefail
go get github.com/google/go-jsonnet/cmd/jsonnet

unrecognized import path "gopkg.in/yaml.v2": reading https://gopkg.in/yaml.v2?go-get=1: 502 Bad Gateway
	server response: Cannot obtain refs from GitHub: cannot talk to GitHub: Get https://github.com/go-yaml/yaml.git/info/refs?service=git-upload-pack: net/http: request canceled (Client.Timeout exceeded while awaiting headers)

Exited with code exit status 1
CircleCI received exit code 1
```

File: docs/source/starter/introduction_guide.rst [hunk 1]
```diff
@@ -127,9 +128,45 @@ Now we add the training_step which has all our training loop logic
...
+            return Adam(self.generator(), lr=1e-3), Adam(self.discriminator(), lr=1e-3)
+
 Data
 ----
 
+Install torchvision library.
+
+.. code-block:: shell
+
+    pip install torchvision
```
Comment (akihironitta) [inline docs/source/starter/introduction_guide.rst:169]: Reviewers: Placing this installation guide here could be distracting to some users maybe? Instead, should there be an installation guide for all extra packages at the very top of the page?

e.g. 

```
Install extra packages used in this tutorial:

.. code-block:: shell

    pip install torchvision transformers ...

```

File: docs/source/starter/introduction_guide.rst [hunk 2]
```diff
@@ -431,6 +439,9 @@ For clarity, we'll recall that the full LightningModule now looks like this.
             loss = F.nll_loss(logits, y)
             return loss
 
+        def configure_optimizers(self):
+            return Adam(self.parameters(), lr=1e-3)
+
```
Comment (akihironitta) [inline docs/source/starter/introduction_guide.rst:444]: Great fix :)

File: docs/source/starter/introduction_guide.rst [hunk 3]
```diff
@@ -170,6 +207,9 @@ Pass in the dataloaders to the `.fit()` function.
 
 .. code-block:: python
 
+    from pytorch_lightning import Trainer
+
+
```
Comment (akihironitta) [inline docs/source/starter/introduction_guide.rst:212]: Reviewers: Instead of having import statements scattered in the page, should we have a single place to import all functions/classes at the very top of this page maybe? Similarly to https://pytorch-lightning.readthedocs.io/en/latest/starter/new-project.html#step-0-install-pytorch-lightning

**Review by akihironitta — APPROVED** [review]
@os1ma Thank you for your feedback on the introduction guide!

File: docs/source/starter/introduction_guide.rst [hunk 4]
```diff
@@ -127,9 +128,45 @@ Now we add the training_step which has all our training loop logic
             loss = F.nll_loss(logits, y)
             return loss
 
+Optimizer
+---------
+
+Next we choose what optimizer to use for training our system.
```
Comment (tchaton) [inline docs/source/starter/introduction_guide.rst:134]:
```suggestion
Next, we choose which optimizer to use for training our model.
```

Comment (tchaton) [reply docs/source/starter/introduction_guide.rst:134]: system isn't defined nor introduced.

File: docs/source/starter/introduction_guide.rst [hunk 5]
```diff
@@ -127,9 +128,45 @@ Now we add the training_step which has all our training loop logic
...
+In PyTorch we do it as follows:
+
+.. code-block:: python
+
+    from torch.optim import Adam
+
+    optimizer = Adam(LitMNIST().parameters(), lr=1e-3)
+
+
+In Lightning we do the same but organize it under the :func:`~pytorch_lightning.core.LightningModule.configure_optimizers` method.
```
Comment (tchaton) [inline docs/source/starter/introduction_guide.rst:144]:
```suggestion
In Lightning, this code is organized within the :func:`~pytorch_lightning.core.LightningModule.configure_optimizers` method of the LightningModule.
```

File: docs/source/starter/introduction_guide.rst [hunk 6]
```diff
@@ -127,9 +128,45 @@ Now we add the training_step which has all our training loop logic
...
+
+.. testcode::
+
+    class LitMNIST(LightningModule):
+        def configure_optimizers(self):
+            return Adam(self.parameters(), lr=1e-3)
+
+.. note:: The LightningModule itself has the parameters, so pass in self.parameters()
```
Comment (tchaton) [inline docs/source/starter/introduction_guide.rst:152]:
```suggestion
.. note:: The LightningModule is a subclassing :class:`~torch.nn.Module`  and therefore, you can access its children parameters with ``self.parameters()``.
```

File: docs/source/starter/introduction_guide.rst [hunk 7]
```diff
@@ -127,9 +128,45 @@ Now we add the training_step which has all our training loop logic
...
+
+However, if you have multiple optimizers use the matching parameters
```
Comment (tchaton) [inline docs/source/starter/introduction_guide.rst:154]:
```suggestion
If you have multiple optimizers, you can configure them as follows:
```

**Comment by os1ma** [comment]
Hi @tchaton.
Thank you for your review! I applied suggestions.

In addition to the review, I noticed that doctest fails by my change. I'll fix that too.

**Comment by os1ma** [comment]
I added a commit to fix the missing testoutput to prevent doctest from failing.

File: docs/source/starter/introduction_guide.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source/starter/introduction_guide.rst:174]: Agree this may be distracting and breaks the flow. Pretty sure this can be left out entirely. If the user does not know how to install this Python package while seeing an import error, they are probably not ready to use Lightning at this point anyway.

File: docs/source/starter/introduction_guide.rst (hunk 3 above)
Comment (awaelchli) [reply docs/source/starter/introduction_guide.rst:217]: Agree. The imports should be done once in the first example where they fit in and then not be repeated. If a new section appears that needs new imports, we add them.

The individual code snippets are anyway not runnable standalone and have to be taken with the context of the whole page.

File: docs/source/starter/introduction_guide.rst (hunk 1 above)
Comment (akihironitta) [reply docs/source/starter/introduction_guide.rst:174]: > Pretty sure this can be left out entirely.

That's what I thought, too. If we do this, we should do this to the entire docs for consistency, which can be quite redundant and less readable...

@os1ma Could you revert the change here at least for now?

Comment (os1ma) [reply docs/source/starter/introduction_guide.rst:174]: Thanks for the review!
I agree that this `pip install` explanation is not necessary for Lightning users. I will remove this part.

File: docs/source/starter/introduction_guide.rst (hunk 3 above)
Comment (os1ma) [reply docs/source/starter/introduction_guide.rst:217]: Thanks for this comment too!
I will move the import to the other imports location at the top of the section.

**Comment by os1ma** [comment]
@awaelchli @akihironitta 

Thank you very much for the review!

Fixed as follows!

* Remove unnecessary `pip install torchvision ` description
* Move the import to the other imports location at the top of the section

File: docs/source/starter/introduction_guide.rst [hunk 8]
```diff
@@ -127,10 +133,40 @@ Now we add the training_step which has all our training loop logic
             loss = F.nll_loss(logits, y)
             return loss
 
+Optimizer
```
Comment (Borda) [inline docs/source/starter/introduction_guide.rst:136]: why is this moved? :rabbit:

**Review by justusschock — APPROVED** [review]

File: docs/source/starter/introduction_guide.rst (hunk 8 above)
Comment (os1ma) [reply docs/source/starter/introduction_guide.rst:136]: Hi @Borda. Thanks for the review!

If the original order that "Optimizer" section is after the "Data" section, the code in the "Data" section fails to execute. This is because the required configure_optimizers method is defined in the "Optimizer" section after the "Data" section.

So it may be good that we swap the order with "Data" section and "Optimizer" section to avoid the problem.

If it is not reasonable, I'll be happy to fix the change!
